File: src/test/java/com/github/davidcarboni/cryptolite/RandomTest.java
package com.github.davidcarboni.cryptolite;

import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Field;
import java.security.SecureRandom;

import static org.junit.Assert.*;

/**
 * Test for {@link Random}.
 *
 * @author David Carboni
 */
public class RandomTest {

    /**
     * Clears the cached instance.
     *
     * @throws NoSuchFieldException   {@link NoSuchFieldException}.
     * @throws IllegalAccessException {@link IllegalAccessException}.
     */
    @Before
    public void setUp() throws NoSuchFieldException, IllegalAccessException {
        Field field = Random.class.getDeclaredField("secureRandom");
        field.setAccessible(true);
        field.set(Random.class, null);
    }

    /**
     * Test method for {@link com.github.davidcarboni.cryptolite.Random#getInstance()}. Checks that
     * {@link Random#getInstance()} returns the same instance on every call, avoiding initialising a
     * new instance every time.
     */
    @Test
    public void testGetInstance() {

        // Given
        SecureRandom firstCall;
        SecureRandom secondCall;

        // When
        firstCall = Random.getInstance();
        secondCall = Random.getInstance();

        // Then
        assertSame(firstCall, secondCall);
    }

    /**
     * Checks that generating a random byte array returns the expected number of bytes.
     */
    @Test
    public void testByteArray() {

        // Given
        int length = 20;

        // When
        byte[] data = Random.byteArray(length);

        // Then
        assertEquals(length, data.length);
    }

    /**
     * Checks that the number of bits in the returned ID is the same as specified by {@link Random#TOKEN_BITS}.
     */
    @Test
    public void testTokenLength() {

        // When
        // We generate a token
        String token = Random.token();

        // Then
        // It should be of the expected length
        byte[] tokenBytes = ByteArray.fromHexString(token);
        assertEquals(Random.TOKEN_BITS, tokenBytes.length * 8);
    }

    /**
     * Checks that the number of bytes in a returned salt value matches the length specified in
     * {@link Random#SALT_BYTES}.
     */
    @Test
    public void testSaltLength() {

        // When
        // We generate a salt
        String salt = Random.salt();

        // Then
        // It should be of the expected length
        byte[] salt_bytes = ByteArray.fromBase64String(salt);
        assertEquals(Random.SALT_BYTES, salt_bytes.length);
    }

    /**
     * Verifies that a random input stream provides the expected amout of input.
     *
     * @throws IOException .
     */
    @Test
    public void testInputStream() throws IOException {

        // Given
        int length = 1025;
        InputStream inputStream = Random.inputStream(length);

        // When
        int count = 0;
        while (inputStream.read() != -1) {
            count++;
        }

        // Then
        assertEquals(length, count);
    }

    /**
     * Checks the number of characters in the returned password matches the specified length of the password.
     */
    @Test
    public void testPasswordLength() {

        // Given
        String password;
        final int maxLength = 100;

        for (int length = 1; length < maxLength; length++) {

            // When
            password = Random.password(length);

            // Then
            assertEquals(length, password.length());
        }
    }

    /**
     * Test the general randomness of token generation.
     * <p>
     * If this test fails, consider yourself astoundingly lucky.. or check the code is really producing random numbers.
     */
    @Test
    public void testRandomnessOfTokens() {

        final int iterations = 1000;
        for (int i = 0; i < iterations; i++) {

            // When
            String id1 = Random.token();
            String id2 = Random.token();

            // Then
            assertNotEquals(id1, id2);
        }
    }

    /**
     * Test the general randomness of salt generation.
     * <p>
     * If this test fails, consider yourself astoundingly lucky.. or check the code is really producing random numbers.
     */
    @Test
    public void testRandomnessOfSalt() {

        final int iterations = 1000;
        for (int i = 0; i < iterations; i++) {

            // When
            String salt1 = Random.salt();
            String salt2 = Random.salt();

            // Then
            assertNotEquals(salt1, salt2);
        }
    }

    /**
     * Test the general randomness of password generation.
     * <p>
     * If this test fails, consider yourself astoundingly lucky.. or check the code is really producing random numbers.
     */
    @Test
    public void testRandomnessOfPasswords() {

        final int iterations = 1000;
        final int passwordSize = 8;
        for (int i = 0; i < iterations; i++) {

            // When
            String password1 = Random.password(passwordSize);
            String password2 = Random.password(passwordSize);

            // Then
            assertNotEquals(password1, password2);
        }
    }

}

File: random/random_test.go
package random

import (
	"regexp"
	"testing"

	"github.com/davidcarboni/cryptolite/bytearray"
)

// Checks that generating a random byte array returns the expected number of bytes.
func TestByteArray(t *testing.T) {

	// Given
	length := 20

	// When
	// Some random bytes
	randomBytes := ByteArray(length)

	// Then
	// Check we got what we expected
	if length != len(randomBytes) {
		t.Error("Unexpected random byte lenth.")
	}
}

// Checks that the number of bits in the returned ID is the same as specified by TokenBits.
func testTokenLength(t *testing.T) {

	// When
	// We generate a token
	token := Token()

	// Then
	// It should be of the expected length
	tokenBytes, err := bytearray.FromHexString(token)
	if err != nil {
		t.Error(err)
	}
	if TokenBits != len(tokenBytes)*8 {
		t.Error("Unexpected token bit-length")
	}
}

// Checks that the number of bytes in a returned salt value matches the length specified in SaltBytes.
func TestSaltLength(t *testing.T) {

	// When
	// We generate a salt
	salt := Salt()

	// Then
	// It should be of the expected length
	saltBytes, err := bytearray.FromBase64String(salt)
	if err != nil {
		t.Error(err)
	}
	if SaltBytes != len(saltBytes) {
		t.Error("Unexpected salt byte-length")
	}
}

// Checks the number of characters and the content of the returned password matches the expected content.
func TestPassword(t *testing.T) {

	// Given
	var password string
	maxLength := 100

	for length := 1; length < maxLength; length++ {

		// When
		password = Password(length)

		// Then
		if length != len(password) {
			t.Error("Unexpected password length")
		}
		match, _ := regexp.MatchString("[A-Za-z0-9]+", password)
		if !match {
			t.Error("Unexpected password content")
		}
	}
}

// Test the general randomness of token generation.
//If this test fails, consider yourself astoundingly lucky.. or check the code is really producing random numbers.
func TestRandomnessOfTokens(t *testing.T) {

	iterations := 1000
	for i := 0; i < iterations; i++ {

		// When
		token1 := Token()
		token2 := Token()

		// Then
		if token1 == token2 {
			t.Error("Got identical tokens.")
		}
	}
}

// Test the general randomness of salt generation.
//If this test fails, consider yourself astoundingly lucky.. or check the code is really producing random numbers.
func TestRandomnessOfSalt(t *testing.T) {

	iterations := 1000
	for i := 0; i < iterations; i++ {

		// When
		salt1 := Salt()
		salt2 := Salt()

		// Then
		if salt1 == salt2 {
			t.Error("Got identical salts.")
		}
	}
}

// Test the general randomness of password generation.
// If this test fails, consider yourself astoundingly lucky.. or check the code is really producing random numbers.
func TestRandomnessOfPasswords(t *testing.T) {

	iterations := 1000
	passwordSize := 8
	for i := 0; i < iterations; i++ {

		// When
		password1 := Password(passwordSize)
		password2 := Password(passwordSize)

		// Then
		if password1 == password2 {
			t.Error("Got identical passwords.")
		}
	}
}
